File: Chapter6/lab6-6/Python/sender.py
import configparser
import socket

if __name__ == "__main__":
    config = configparser.ConfigParser()
    config.read('./config.ini')
    mss = int(config["config"]["MSS"])
    print("MSS = {}".format(mss))
    threshold = int(config["config"]["Threshold"])
    print("Initial threshold = {}".format(threshold))
    triACKRound = int(config["config"]["TriACKRound"])
    timeoutRound = int(config["config"]["TimeoutRound"])
    endRound = int(config["config"]["EndRound"])
    
    conWindow = 1
    with socket.socket(socket.AF_INET) as s:
        s.bind(("127.0.0.1", 6677))
        s.connect(("127.0.0.1", 7788))
        round = 0
        flag = '*2'
        while round < endRound:
            round += 1

            if conWindow >= threshold:
                flag = '+1'

            if round % triACKRound == 0:
                print(" Triple ACK ")
                round -= 1
                print("sending: " + str(round))
                print("round: " + str(round))
                print("window size" + str(conWindow))
                s.send(bytes(str(round), encoding='utf-8'))
                round += 1
                threshold = conWindow / 2
                conWindow = threshold + 3
                continue
            
            if round % timeoutRound == 0:
                print(" Time out ")
                threshold = conWindow / 2
                conWindow = 1
                flag = '*2'

            print("sending: " + str(round))
            print("round: " + str(round))
            print("window size" + str(conWindow))
            s.send(bytes(str(round), encoding='utf-8'))

            if flag == '*2':
                conWindow *= 2
            elif flag == '+1':
                conWindow += 1

            

File: Chapter6/lab6-6/Java/Sender.java
package network.chapter6.lab6;

public class Sender 
{
	private int _mss;
	private int _threshold;
	private int _triAckRound;
	private int _timeoutRound;
	private int _endRound;
	private Receiver _receiver;
	private byte[] arr;
	
	Sender(int mss, int threshold, int triAckRound, int timeoutRound, int endRound, Receiver receiver)
	{
		_mss = mss;
		_threshold = threshold;
		_triAckRound = triAckRound;
		_timeoutRound = timeoutRound;
		_endRound = endRound;
		_receiver = receiver;
		arr = new byte[mss];
		for (int i = 0; i < mss; i++)
		{
			arr[i] = 'a';
		}
	}
	
	void start()
	{
		System.out.println("MSS: " + _mss);
		System.out.println("Initial threshold: " + _threshold);
		int cwnd = 1;
		
		for (int round = 1; round <= _endRound; round++)
		{
			System.out.println("---------------------------");
			System.out.println("Round: " + round);
			System.out.println("cwnd: " + cwnd);
			System.out.println("threshold: " + _threshold);
			int counter = 0;
			for (int j = 0; j < cwnd; j++)
			{
				_receiver.receive(arr);
				counter += arr.length;
			}
			System.out.println("Send data: a(repeat " + counter + " times)");
			
			if (round == _triAckRound)
			{
				cwnd /= 2;
				_threshold = cwnd;
			}
			else if (round == _timeoutRound)
			{
				_threshold = cwnd / 2;
				cwnd = 1;
			}
			else
			{
				if (cwnd < _threshold)
				{
					cwnd *= 2;
				}
				else
				{
					cwnd++;
				}
			}
			
		}
	}
}
